src/main/java/graphql/execution/reactive/CompletionStageSubscriber.java

```diff
@@ -0,0 +1,201 @@
+package graphql.execution.reactive;
+
+import graphql.Internal;
+import graphql.util.LockKit;
+import org.jetbrains.annotations.NotNull;
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+
+import java.util.ArrayDeque;
+import java.util.Queue;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.BiConsumer;
+import java.util.function.Function;
+
+/**
+ * This subscriber can be used to map between a {@link org.reactivestreams.Publisher} of U
+ * elements and map them into {@link CompletionStage} of D promises.
+ *
+ * @param <U> published upstream elements
+ * @param <D> mapped downstream values
+ */
+@Internal
+public class CompletionStageSubscriber<U, D> implements Subscriber<U> {
```

Comment:
This was split out from the peer Publisher code as its own file.  Its complicated enough to deserve that.